### PR TITLE
Time dimensions

### PR DESCRIPTION
**Pull request by jgoizueta**
See #1037 (WIP)

### CODE REVIEW

**Comment by jgoizueta** [comment]
Hey @dgaubert  it would be nice that someone starts to review this, the CARTO VL front to expose this ain't finished yet but I don't forsee relevant changes here. There's some documentation [here](https://github.com/CartoDB/Windshaft-cartodb/blob/a4dfc09c719b3f78e08ed20d62c93f200a3394e4/docs/aggregation.md#dimensions):

File: docs/aggregation.md [hunk 1]
```diff
@@ -134,6 +134,61 @@ of the original dataset applying three different aggregate functions.
...
+### `dimensions`
+
+By default aggregated data is grouped only by the the spatial discretization into grid cells.
+The `dimensions` parameters permits to define additional aggregation dimensions. Currently only date (time, timestamp) columns can be used as dimensions. Just like the spatial position is discretized using a grid, for every
+additional dimension, a grouping into discrete periods of time must be defined.
+
+```json
+{
+    "dimensions": {
+        "time_in_days": { // Name of the resulting dimension column
```
Comment (jgoizueta) [inline docs/aggregation.md]: 🤔 I should either remove the comments or change it to javascript

Comment (dgaubert) [reply docs/aggregation.md]: use `js` instead.

File: lib/cartodb/backends/layer-stats/mapnik-layer-stats.js [hunk 2]
```diff
@@ -178,6 +201,18 @@ function _columnStats(ctx, columns) {
                         stats[name][fn] = res.rows[0][`${name}_${fn}`];
                     });
                 });
+                Object.keys(dimensionsInfo).forEach(name => {
+                    // Temporalily place dimensions info in stats.columns.__dimensions
+                    stats.__dimensions = stats.__dimensions || {};
```
Comment (dgaubert) [inline lib/cartodb/backends/layer-stats/mapnik-layer-stats.js]: Why do we need `__dimensions` as temporally placeholder?. Are we hacking a bit `stats` to prepare dimensions statistics later?

File: lib/cartodb/backends/layer-stats/mapnik-layer-stats.js [hunk 3]
```diff
@@ -137,35 +142,53 @@ function _sample(ctx, numRows) {
     return Promise.resolve();
 }
 
-function _columnStats(ctx, columns) {
+function _columnStats(ctx, columns, dimensions) {
     if (!columns) {
         return Promise.resolve();
     }
-    if (ctx.metaOptions.columnStats) {
+    if (ctx.metaOptions.columnStats || ctx.metaOptions.dimensions) {
```
Comment (dgaubert) [inline lib/cartodb/backends/layer-stats/mapnik-layer-stats.js]: I'd love to see some refactor here, it's getting hard to understand what's happening. I think by extracting some common functionality and then splitting conditionals would be enough.

File: lib/cartodb/backends/layer-stats/mapnik-layer-stats.js (hunk 2 above)
Comment (jgoizueta) [reply lib/cartodb/backends/layer-stats/mapnik-layer-stats.js]: I got into this ugly mess ad I couldn't find a better way out: Here we initially collected info about columns only and all the results are gathered at the end of this function and merged into a `columns` property:
```
return Promise.all(queries).then(results => ({ columns: mergeColumns(results) }));
```
But now I also need to put some of the results in a `dimensions` property, so I use `__dimensions` in hope no column will be ever named like that 🤞 and later, where all the queries are executed it is moved to where it should be.

I'll give it another thought with a fresher mind tomorrow 😅

**Pull request merged**